Utilities/build-using-self

```diff
@@ -1,62 +1,163 @@
-    echo "Current working directory is ${PWD}"
-    echo "Bootstrapping with the XCBuild codepath..."
-    ${root_dir}/Utilities/bootstrap \
-      build \
-      --release \
-      --verbose \
-      --cross-compile-hosts macosx-arm64 \
-      --skip-cmake-bootstrap \
-      --swift-build-path "${SWIFTPM_BIN_DIR}/swift-build"
-fi
+
+REPO_ROOT_PATH = pathlib.Path(__file__).parent.parent.resolve()
+
+
+def get_arguments() -> argparse.Namespace:
+    parser = argparse.ArgumentParser(
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter
+    )
+
+    parser.add_argument(
+        "-v",
+        "--verbose",
+        dest="is_verbose",
+        action="store_true",
+        help="When set, prints verbose information.",
+    )
+    parser.add_argument(
+        "-c",
+        "--configuration",
+        type=Configuration,
+        dest="config",
+        default="debug",
+        choices=[e.value for e in Configuration],
+        help="The configuraiton to use.",
+    )
+
+    args = parser.parse_args()
+    return args
+
+
+def log_environment() -> None:
+    logging.info("Environment Variables")
+    for key, value in sorted(os.environ.items()):
+        logging.info("  --> %s=%r", key, value)
+
+
+def get_swiftpm_bin_dir(config: Configuration) -> pathlib.Path:
+    swiftpm_bin_dir = pathlib.Path(
+        call_output(["swift", "build", "--configuration", config, "--show-bin-path"])
+    )
+    logging.info("SwiftPM BIN DIR: %s", swiftpm_bin_dir)
+    return swiftpm_bin_dir
+
+
+def is_on_darwin() -> bool:
+    return platform.uname().system == "Darwin"
+
+
+def set_environment(*, config: Configuration) -> None:
+    os.environ["SWIFTCI_IS_SELF_HOSTED"] = "1"
+
+    # Set the SWIFTPM_BIN_DIR path
+    os.environ["SWIFTPM_BIN_DIR"] = str(get_swiftpm_bin_dir(config=config))
+
+    # Ensure SDKROOT is configure
+    if is_on_darwin():
+        sdk_root = call_output(["xcrun", "--show-sdk-path", "--sdk", "macosx"])
+        logging.debug("macos sdk root = %r", sdk_root)
+        os.environ["SDKROOT"] = sdk_root
+    log_environment()
+
+
+def run_bootstrap(swiftpm_bin_dir: pathlib.Path) -> None:
+    if is_on_darwin():
+        logging.info("Current working directory is %s", pathlib.Path.cwd())
+        logging.info("Bootstrapping with the XCBuild codepath...")
+        call(
+            [
+                REPO_ROOT_PATH / "Utilities" / "bootstrap",
+                "build",
+                "--release",
+                "--verbose",
+                "--cross-compile-hosts",
+                "macosx-arm64",
+                "--skip-cmake-bootstrap",
+                "--swift-build-path",
+                (swiftpm_bin_dir / "swift-build").resolve(),
+            ],
+        )
+
+
+def main() -> None:
+    args = get_arguments()
+    logging.getLogger().setLevel(logging.DEBUG if args.is_verbose else logging.INFO)
+    logging.debug("Args: %r", args)
+
+    with change_directory(REPO_ROOT_PATH):
+        set_environment(config=args.config)
```

Comment:
suggestion (non-blocking): pass the swift binary directory as an argument to `set_environment(...)`

Reply:
done